### ISSUE TITLE
bad file descriptor endless printout - took the server down with full disk

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#1525]
server logs are full of the below repeated over and over again:

```
>>>>java.io.IOException: Bad file descriptor
        at java.io.FileInputStream.readBytes(Native Method)
        at java.io.FileInputStream.read(FileInputStream.java:255)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.fill(BufferedReader.java:161)
        at java.io.BufferedReader.readLine(BufferedReader.java:324)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at games.strategy.engine.lobby.server.headless.HeadlessLobbyConsole.printEvalLoop(HeadlessLobbyConsole.java:69)
        at games.strategy.engine.lobby.server.headless.HeadlessLobbyConsole.lambda$start$263(HeadlessLobbyConsole.java:59)
        at java.lang.Thread.run(Thread.java:745)
```

```
danv@tripleawarclub:~$ wc -l nohup.out
431594515 nohup.out
```

**Comment by RoiEXLab** [comment]
@DanVanAtta #1527 should fix this issue, you'll need to test this out, but it's a start

**Comment by ssoloff** [comment]
Quick drive-by comment here...

Every time I've run into a `Bad file descriptor` exception, it's been because someone (usually me) had inadvertently closed the underlying `InputStream` from under me.  I took a quick look at `LobbyServer` and `HeadlessLobbyConsole`, and I didn't see any explicit or implicit closing of the input stream.  In fact, there's a comment (referring to the `out` stream) to the effect of "don't close the stream because it will ultimately close `System.out`."

So, it **sounds** like `stdin` was closed somehow, but after my quick reading of the code, I don't see how it could be caused by TripleA.  My *nix system skills aren't too deep, so I'm not sure if this is even possible, but could there have been something fatal that happened on the server that caused the process' `stdin` to be closed (but left the process running)?  Was there anything interesting in either the TriplaA log or the system log prior to the first instance of the above exception?

**Comment by RoiEXLab** [comment]
@ssoloff Absolutely correct. The problem isn't that we are closing stdin, the problem is that nohup (the application we use for running the bots in the background) seems to be closing that resource

**Comment by RoiEXLab** [comment]
I tested this, with a test jar and it seems to be working this way...

**Comment by ssoloff** [comment]
Thanks, @RoiEXLab.  It's been such a long time since I've used `nohup` that I glossed over that detail in the log file name at the top of the ticket. ☹️ So, at least that makes sense as to who could possibly be closing `stdin`...

**Comment by RoiEXLab** [comment]
@ssoloff No Problem 😄

**Comment by ssoloff** [comment]
(Please ignore this comment until the TL;DR.  I'm just adding it to document some stuff I learned in case I ever want to refer to it in the future, as I found this problem interesting.)

For some reason, I assumed the above exception started to appear at some time after the server had started (i.e. the `BufferedReader#readLine()` call succeeded for a period of time and only started to fail at some point in the future after `stdin` was closed).  Apparently, `readLine()` fails right from the start when the server is run via `nohup`.

Looking at the [code](http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/nohup.c) for GNU `nohup`, it's immediately clear why this happens.  If `stdin` is a terminal (as opposed to having possibly being redirected from a file), `nohup` redirects the child process' `stdin` to an "unreadable file," which is basically `/dev/null` opened in write-only mode.  Thus, any attempt to read from `stdin` will result in a `Bad file descriptor` exception.

Now I understand why @RoiEXLab's change fixes the problem.  `Scanner#hasNextLine()` properly handles the case when the input stream would normally throw an exception (this is buried in `Scanner#readInput()`) and instead returns `false`, upon which the `HeadlessLobbyConsole` thread exits.

**TL;DR**

Another possible solution could be to not even start the console when `nohup` is used by specifying `triplea.lobby.console=false` on the command line when starting the server.  Actually, it looks like this is the default if this argument is not specified, so someone must be explicitly specifying `triplea.lobby.console=true`.  I think the culprit is *.travis/prerelease_lobby_install*, which contains the following line:

```
nohup java -server  -Xmx192m -classpath bin/triplea.jar:lib/derby-10.10.1.1.jar -Dtriplea.lobby.port=33333 -Dtriplea.lobby.console=true games.strategy.engine.lobby.server.LobbyServer &
```

As implied above, when using `nohup`, the console is useless if `stdin` is a terminal, which it is in the above command.  (Note that I think @RoiEXLab's fix is still a good change, as it properly terminates the input loop upon EOS, which is always a Good Thing.)

**Comment by DanVanAtta** [comment]
No code was changed, this has always been a problem @ron-murhammer.
The lobby is configured in an interesting way that requires it to have a stdin.
For a host of reasons we have agreed moving forward to remove this.

**Comment by ron-murhammer** [comment]
@DanVanAtta Correct. My understanding is this was caused by changing the commands/scripts we are using to launch the lobby.
